src/python/dxpy/utils/config.py

```diff
@@ -0,0 +1,297 @@
+        except Exception as e:
+            warn(fill("Unexpected error while retrieving session configuration: " + format_exception(e)))
+        return os.path.join(sessions_dir, str(os.getpid()))
+
+    def _read_conf_dir(self, dirname):
+        try:
+            with open(os.path.join(dirname, "environment.json")) as fd:
+                env_vars = json.load(fd)
+        except Exception:
+            env_vars = {}
+
+        for standalone_var in self.STANDALONE_VAR_NAMES:
+            try:
+                with open(os.path.join(dirname, standalone_var)) as fd:
+                    env_vars[standalone_var] = fd.read()
+            except Exception:
+                pass
+        return env_vars
+
+    def __getitem__(self, item):
+        if item not in self.VAR_NAMES:
+            raise KeyError(item)
+        return environ[item]
+
+    def __setitem__(self, key, value):
+        if key not in self.VAR_NAMES:
+            raise KeyError(key)
+        if value is None:
+            value = self.defaults.get(key, "")
+        environ[key] = value
+        if key in self.CORE_VAR_NAMES:
+            self._sync_dxpy_state()
+
+    def __delitem__(self, key):
+        if key not in self.VAR_NAMES:
+            raise KeyError(key)
+        del environ[key]
+        if key in self.CORE_VAR_NAMES:
+            self._sync_dxpy_state()
+
+    def __iter__(self):
+        for item in self.VAR_NAMES:
+            if item in environ:
+                yield item
+
+    def __len__(self):
+        return len(self.VAR_NAMES)
+
+    def __repr__(self):
+        desc = "<{module}.{classname} object at 0x{mem_loc:x}: {data}>"
+        return desc.format(module=self.__module__,
+                           classname=self.__class__.__name__,
+                           mem_loc=id(self),
+                           data=dict(self))
+
+    def write(self, item, value):
+        self[item] = value
+        self.save()
+
+    def save(self):
+        self._write_conf_dir(self._user_conf_dir)
+        self._write_conf_dir(self.get_session_conf_dir())
+
+    def _write_conf_dir(self, conf_dir):
+        try:
+            os.makedirs(conf_dir, 0o700)
+        except OSError:
+            os.chmod(conf_dir, 0o700)
+
+        env_jsonfile_path = os.path.join(conf_dir, "environment.json")
+        # Make sure the file has 600 permissions
+        with _open_for_writing_with_permissions(env_jsonfile_path, 0o600) as fd:
+            json.dump({k: self.get(k, self.defaults.get(k)) for k in self}, fd, indent=4)
+            fd.write("\n")
+
+        for var in self.STANDALONE_VAR_NAMES:
+            # Make sure the file has 600 permissions
+            with _open_for_writing_with_permissions(os.path.join(conf_dir, var), 0o600) as fd:
+                value = self.get(var, self.defaults.get(var, ""))
+                fd.write(value.encode(sys_encoding) if USING_PYTHON2 else value)
+
+        if not os.path.exists(os.path.join(self._user_conf_dir, "unsetenv")):
```

Comment:
I feel that this should be factored out into a separate function, perhaps to be called just once in `save`. It's independent of the values in the configuration, and it shouldn't be written twice (once for each call to `_write_conf_dir`.